### ISSUE TITLE
sanity check failed for ...

### ISSUE THREAD

**Issue by albertnetymk** [mvo5/unattended-upgrades#223]
On logging into my system, I was greeted by the following:

```
2 updates could not be installed automatically. For more details,
see /var/log/unattended-upgrades/unattended-upgrades.log
```
The log doesn't seem to contain anything insightful, so I decided to debug it using `-d`. Unfortunately, it's not obvious to me why the sanity check fails. Anyone can give me some instruction on how I can proceed to debug this?
```
$ unattended-upgrade -d -v --dry-run`
Checking if system is running on battery is skipped. Please install powermgmt-base package to check power status and skip installing updates when the system is running on battery
.
Initial blacklist :
Initial whitelist:
Starting unattended upgrades script
Allowed origins are: o=Debian,a=testing, o=Debian,a=testing-updates
Using (^linux-image-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-headers-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-image-extra-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?
$|^linux-modules-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-modules-extra-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-signed-image-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-
.+)?$|^linux-image-unsigned-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^kfreebsd-image-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^kfreebsd-headers-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9
]+(-.+)?$|^gnumach-image-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^.*-modules-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^.*-kernel-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^lin
ux-backports-modules-.*-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-modules-.*-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-tools-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)
?$|^linux-cloud-tools-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-buildinfo-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-source-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$
|^linux-image-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-headers-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-image-extra-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^lin
ux-modules-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-modules-extra-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-signed-image-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|
^linux-image-unsigned-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^kfreebsd-image-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^kfreebsd-headers-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+
)?$|^gnumach-image-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^.*-modules-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^.*-kernel-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-bac
kports-modules-.*-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-modules-.*-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-tools-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^li
nux-cloud-tools-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-buildinfo-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$|^linux-source-[1-9][0-9]*\.[0-9]+\.[0-9]+-[0-9]+(-.+)?$) rege
xp to find kernel packages
Using (^linux-image-5\.2\.0\-2\-amd64$|^linux-image-5\.2\.0\-2$|^linux-headers-5\.2\.0\-2\-amd64$|^linux-headers-5\.2\.0\-2$|^linux-image-extra-5\.2\.0\-2\-amd64$|^linux-image-ex
tra-5\.2\.0\-2$|^linux-modules-5\.2\.0\-2\-amd64$|^linux-modules-5\.2\.0\-2$|^linux-modules-extra-5\.2\.0\-2\-amd64$|^linux-modules-extra-5\.2\.0\-2$|^linux-signed-image-5\.2\.0\
-2\-amd64$|^linux-signed-image-5\.2\.0\-2$|^linux-image-unsigned-5\.2\.0\-2\-amd64$|^linux-image-unsigned-5\.2\.0\-2$|^kfreebsd-image-5\.2\.0\-2\-amd64$|^kfreebsd-image-5\.2\.0\-
2$|^kfreebsd-headers-5\.2\.0\-2\-amd64$|^kfreebsd-headers-5\.2\.0\-2$|^gnumach-image-5\.2\.0\-2\-amd64$|^gnumach-image-5\.2\.0\-2$|^.*-modules-5\.2\.0\-2\-amd64$|^.*-modules-5\.2
\.0\-2$|^.*-kernel-5\.2\.0\-2\-amd64$|^.*-kernel-5\.2\.0\-2$|^linux-backports-modules-.*-5\.2\.0\-2\-amd64$|^linux-backports-modules-.*-5\.2\.0\-2$|^linux-modules-.*-5\.2\.0\-2\-
amd64$|^linux-modules-.*-5\.2\.0\-2$|^linux-tools-5\.2\.0\-2\-amd64$|^linux-tools-5\.2\.0\-2$|^linux-cloud-tools-5\.2\.0\-2\-amd64$|^linux-cloud-tools-5\.2\.0\-2$|^linux-buildinf
o-5\.2\.0\-2\-amd64$|^linux-buildinfo-5\.2\.0\-2$|^linux-source-5\.2\.0\-2\-amd64$|^linux-source-5\.2\.0\-2$|^linux-image-5\.2\.0\-2\-amd64$|^linux-image-5\.2\.0\-2$|^linux-heade
rs-5\.2\.0\-2\-amd64$|^linux-headers-5\.2\.0\-2$|^linux-image-extra-5\.2\.0\-2\-amd64$|^linux-image-extra-5\.2\.0\-2$|^linux-modules-5\.2\.0\-2\-amd64$|^linux-modules-5\.2\.0\-2$
|^linux-modules-extra-5\.2\.0\-2\-amd64$|^linux-modules-extra-5\.2\.0\-2$|^linux-signed-image-5\.2\.0\-2\-amd64$|^linux-signed-image-5\.2\.0\-2$|^linux-image-unsigned-5\.2\.0\-2\
-amd64$|^linux-image-unsigned-5\.2\.0\-2$|^kfreebsd-image-5\.2\.0\-2\-amd64$|^kfreebsd-image-5\.2\.0\-2$|^kfreebsd-headers-5\.2\.0\-2\-amd64$|^kfreebsd-headers-5\.2\.0\-2$|^gnuma
ch-image-5\.2\.0\-2\-amd64$|^gnumach-image-5\.2\.0\-2$|^.*-modules-5\.2\.0\-2\-amd64$|^.*-modules-5\.2\.0\-2$|^.*-kernel-5\.2\.0\-2\-amd64$|^.*-kernel-5\.2\.0\-2$|^linux-backport
s-modules-.*-5\.2\.0\-2\-amd64$|^linux-backports-modules-.*-5\.2\.0\-2$|^linux-modules-.*-5\.2\.0\-2\-amd64$|^linux-modules-.*-5\.2\.0\-2$|^linux-tools-5\.2\.0\-2\-amd64$|^linux-
tools-5\.2\.0\-2$|^linux-cloud-tools-5\.2\.0\-2\-amd64$|^linux-cloud-tools-5\.2\.0\-2$|^linux-buildinfo-5\.2\.0\-2\-amd64$|^linux-buildinfo-5\.2\.0\-2$|^linux-source-5\.2\.0\-2\-
amd64$|^linux-source-5\.2\.0\-2$) regexp to find running kernel packages
Checking: clang ([<Origin component:'main' archive:'testing' origin:'Debian' label:'Debian' site:'deb.debian.org' isTrusted:True>])
pkg libomp-7-dev now marked delete
sanity check failed for: {'clang-8=1:8.0.1-3+b1', 'llvm-8-dev=1:8.0.1-3+b1', 'libomp5-7=1:7.0.1-9+b1', 'libpfm4=4.10.1+git14-g815ff28-1', 'libomp5-8=1:8.0.1-3+b1', 'libclang1-8=1
:8.0.1-3+b1', 'libclang-common-8-dev=1:8.0.1-3+b1', 'libomp-7-dev=1:7.0.1-9+b1', 'libomp-8-dev=1:8.0.1-3+b1', 'llvm-8-runtime=1:8.0.1-3+b1', 'llvm-8=1:8.0.1-3+b1', 'clang=1:8.0-4
8.2'}
Checking: clang-tools ([<Origin component:'main' archive:'testing' origin:'Debian' label:'Debian' site:'deb.debian.org' isTrusted:True>])
pkg libomp-7-dev now marked delete
sanity check failed for: {'clang-tools-8=1:8.0.1-3+b1', 'clang-8=1:8.0.1-3+b1', 'clang-tools=1:8.0-48.2', 'llvm-8-dev=1:8.0.1-3+b1', 'libomp5-7=1:7.0.1-9+b1', 'libpfm4=4.10.1+git
14-g815ff28-1', 'libomp5-8=1:8.0.1-3+b1', 'libclang1-8=1:8.0.1-3+b1', 'libclang-common-8-dev=1:8.0.1-3+b1', 'libomp-7-dev=1:7.0.1-9+b1', 'libomp-8-dev=1:8.0.1-3+b1', 'llvm-8=1:8.
0.1-3+b1', 'llvm-8-runtime=1:8.0.1-3+b1'}
pkgs that look like they should be upgraded:
Fetched 0 B in 0s (0 B/s)
fetch.run() result: 0
blacklist: []
whitelist: []
Option --dry-run given, *not* performing real actions
Packages that will be upgraded:
InstCount=0 DelCount=0 BrokenCount=0
```

ENV:

```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux bullseye/sid
Release:        testing
Codename:       bullseye
```

**Comment by rbalint** [comment]
The reason seems to be:

```
pkg libomp-7-dev now marked delete
```

U-u does not remove packages to upgrade others to new version.
I agree that this should be explained better in the logs and also should be noted without debugging turned on.

**Comment by albertnetymk** [comment]
I don't understand this part: "U-u does not remove packages to upgrade others to new version." I thought `Remove-Unused-Dependencies` is used for removing those packages (leftover dependencies).

**Comment by rbalint** [comment]
@albertnetymk This is a different case. Sometimes APT needs to remove package A to upgrade package B, when there is a Breaks/Conflicts relationship between them. This upgrade would not be performed by u-u.

**Comment by albertnetymk** [comment]
I see; does this mean `apt-get dist-upgrade` will not perform this upgrade as well? If dist-upgrade would go ahead and perform upgrade, I don't get why u-u is being extraneous cautious.

PS: Did some googling regarding cron+dist-upgrade+auoremove vs u-u, but can't seem to find any significant difference. Some clarification in this regard is highly appreciated.

**Comment by rbalint** [comment]
@albertnetymk I suggest reading documentation further.

**Comment by albertnetymk** [comment]
Another instance of the same error.

```
Checking: libsmbclient ([<Origin component:'main' archive:'testing' origin:'Debian' label:'Debian' site:'deb.debian.org' isTrusted:True>])
  Installing samba-libs as Depends of libsmbclient
    Installing libldb2 as Depends of samba-libs
       Removing: libldb1
    Installing libtracker-sparql-2.0-0 as Depends of samba-libs
    Installing libwbclient0 as Depends of samba-libs
    Installing python3-talloc as Depends of samba-libs
pkg libldb1 now marked delete
sanity check failed for: {'samba-libs=2:4.11.0+dfsg-10', 'python3-talloc=2.3.0-2', 'libldb2=2:2.0.7-3', 'libwbclient0=2:4.11.0+dfsg-10', 'libsmbclient=2:4.11.0+dfsg-10', 'libtracker-sparql-2.0-0=2.1.8-2', 'libldb1=2:1.5.1+really1.4.7-1'}
```

The corresponding source code is at https://github.com/mvo5/unattended-upgrades/blob/master/unattended-upgrade#L949-L951 It's not clear to me why the sanity check fails if a package is marked deleted.

**Comment by rbalint** [comment]
Fixed by #255
